src/main.rs

```diff
@@ -31,81 +33,138 @@ struct Kopium {
+            match self.command {
+                Some(Command::ListCrds) => self.list_crds(crds).await?,
+                Some(Command::Completions { shell }) => self.completions(shell)?,
+                None => self.help()?,
+            }
+        }
 
+        Ok(())
+    }
 
-    if let Some(schema) = data {
-        let mut results = vec![];
-        log::debug!("schema: {}", serde_json::to_string_pretty(&schema)?);
-        analyze(schema, "", &kind, 0, &mut results)?;
+    fn help(&self) -> Result<()> {
+        Self::clap().print_help().map(|_| println!())?;
+        Ok(())
+    }
 
-        if !kopium.hide_prelude {
-            print_prelude(&results);
-        }
+    async fn list_crds(&self, crds: Crds) -> Result<()> {
+        let lp = api::ListParams::default();
+        crds.list(&lp).await?.items.iter().for_each(|crd| {
+            println!("{}", crd.name());
+        });
+        Ok(())
+    }
 
-        for s in results {
-            if s.level == 0 {
-                continue; // ignoring root struct
-            } else {
-                if s.level == 1 && s.name.ends_with("Spec") {
-                    print_derives(&kopium.derive);
-                    println!(
-                        r#"#[kube(group = "{}", version = "{}", kind = "{}", plural = "{}")]"#,
-                        group, version, kind, plural
-                    );
-                    if scope == "Namespaced" {
-                        println!(r#"#[kube(namespaced)]"#);
-                    }
-                    // don't support grabbing original schema atm so disable schemas:
-                    // (we coerce IntToString to String anyway so it wont match anyway)
-                    println!(r#"#[kube(schema = "disabled")]"#);
-                    println!("pub struct {} {{", s.name);
+    fn completions(&self, shell: clap::Shell) -> Result<()> {
+        use std::io::Write;
+        let mut stdout = std::io::stdout();
+        Self::clap().gen_completions_to("kopium", shell, &mut stdout);
+        writeln!(stdout)?;
+        Ok(())
+    }
+
+    async fn generate(&self, crds: Crds, name: &str) -> Result<()> {
+        let crd = crds.get(name).await?;
+
+        let version = find_crd_version(&crd, self.api_version.as_deref())?;
+        let data = version
+            .schema
+            .as_ref()
+            .and_then(|schema| schema.open_api_v3_schema.clone());
+        let version = version.name.clone();
+
+        let kind = crd.spec.names.kind;
+        let plural = crd.spec.names.plural;
+        let group = crd.spec.group;
+        let scope = crd.spec.scope;
+
+
+        if let Some(schema) = data {
+            let mut results = vec![];
+            log::debug!("schema: {}", serde_json::to_string_pretty(&schema)?);
+            analyze(schema, "", &kind, 0, &mut results)?;
+
+            if !self.hide_prelude {
+                print_prelude(&results);
+            }
+
+            for s in results {
+                if s.level == 0 {
```

Comment:
it might be worth extracting the print out into its separate function at this point.
something like `print_structs(args: &Kopium, structs: &[OutputStruct])`

Reply:
oh, actually, you'd need a bunch of variables from main to do this.. (kind, plural, group, scope, version), so it would be a bit awkward.